### DISCUSSION TITLE
Okta saml Integration

### DISCUSSION THREAD

**Discussion by VAIBHAV631** [fastapi/fastapi#9223] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi.responses import RedirectResponse
import os
from fastapi import Request,Response
from starlette.responses import HTMLResponse, RedirectResponse
from onelogin.saml2.auth import OneLogin_Saml2_Auth
from onelogin.saml2.utils import OneLogin_Saml2_Utils
router = APIRouter()
async def prepare_from_fastapi_request(request, debug=False):
    form_data = await request.form()
    rv = {
    # "https": request.is_secure,
    "http_host": request.client.host,
    "server_port": request.url.port,
    "script_name": request.url.path,
    "post_data": { },
    "get_data": { }
  }
    print(rv)
    if (request.query_params):
        rv["get_data"] = request.query_params,
    if "SAMLResponse" in form_data:
        SAMLResponse = form_data["SAMLResponse"]
        rv["post_data"]["SAMLResponse"] = SAMLResponse
    if "RelayState" in form_data:
        RelayState = form_data["RelayState"]
        rv["post_data"]["RelayState"] = RelayState
    return rv

# okta credientials
saml_settings = {
  "strict": False, # can set to True to see problems such as Time skew/drift
  "debug": True,
  "idp": {
    "entityId": "http://www.okta.com/exe32ssdssddddddddddddddddk8ffcudcsssssssssVCaS2Ed5d7",
    "singleSignOnService": {
      "url": "https://dev-464664646.okta.com/app/dev-138041ds773_oktafastapi_1/exe32ssdssddddddddddddddddk8ffcudcsssssssssVCaS2Ed5d7/sso/saml",
      "binding": "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect"
    },
    "x509cert": "fffffffffffff"
  },
  "sp": {
    "entityId": "http://localhost:8000/saml2/login",
    "assertionConsumerService": {
      "url": "http://localhost/api/v1/saml2/callback",
      "binding": "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
    },
    "x509cert": "fffffffffffffffffff",
  }
}

#SAML-Login-SSO-Client Crendedatiols
@router.get('/saml2/login',response_class=RedirectResponse)
async def saml_login(request: Request):
  req = await prepare_from_fastapi_request(request)
  # print(req)
  auth = OneLogin_Saml2_Auth(req, saml_settings)
  # print(auth)
  callback_url = auth.login()
  print((f"Hello, {auth.get_nameid()}"))
  # saml_call = await saml_callback()

  return RedirectResponse(url=callback_url)

@router.post("/saml2/callback")
async def saml_callback(request: Request):
    # auth = OneLogin_Saml2_Auth(request, saml_settings)
    # auth.process_response()
    req = await prepare_from_fastapi_request(request, True)
    auth = OneLogin_Saml2_Auth(req, saml_settings)
    auth.process_response()
    errors = auth.get_errors()
    if not errors:
        if auth.is_authenticated():
            return RedirectResponse("http://localhost:3000")

        else:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND,detail=f"error")
    else:
        # if settings.DEBUG:
        #     print(auth.get_last_error_reason())
        return RedirectResponse("Error when processing SAML Response: %s" % (', '.join(errors)))
```

### Description

@tiangolo
I am implementing okta saml integration , the first route is /saml2/login and second route is 
/saml2/callback  , areas of improvement in the second route is , authenticate user must redirect to home page.
current getting {"detail":"Not Found"}.

please suggest the solution for  same.

### Operating System

Windows

### FastAPI Version

Name: fastapi Version: 0.89.1

### Python Version

Python 3.10.4

**Comment by yashp280793** [comment]
Did you get any solution on this? I am also implementing the same but facing the redirection issue.

@tiangolo Any direction will be very helpful. thank you.

**Comment by yashp280793** [comment]
@VAIBHAV631 I got the solution of this.

If you do not mention status_code = 302 with RedirectResponse, it will initiate POST request.  so add that as below:

return RedirectResponse("http://localhost:3000", status_code=302)

Along with this, you will have to create and set the token to the same after the redirection as below:

response.set_cookie(key="token", value="value of token", domain:"localhost", samesite="none",secure=True)

Try and test the same. It should work.

**Comment by mostopalove** [comment]
Hey, @VAIBHAV631 ! I faced the same task. 
How did you deal with it after all?
And are you able to get user details in the backend?
